lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs

```diff
@@ -1375,6 +1492,205 @@ makeChangeForCoin
+--      change bundle completely, we just leave it as an empty entry
+--      (effectively "pad with zeros").
+--   2. By removing from the smallest bundle, we maintain the ascending partial
+--      ordering of the change bundles.
+--   3. By removing from the smallest bundles, we remove the "least useful"
+--      bundles, maximizing the overall usefulness of our bundles.
+
+-- | Adds a minted value to the change map. This function presumes that the
+-- change map is partially-ordered in ascending order (i.e. the largest
+-- element is at the end of the list).
+--
+-- Properties:
+-- 1. wrt to the total value of assets in the changeMaps, calling this
+-- function is the same as adding the mint value from the change maps:
+--    F.fold changeMaps <> TokenMap.singleton assetId qty
+--          == F.fold (f (assetId, qty) changeMaps)
+-- 2. length is conserved
+-- 3. order is preserved (i.e., the result is in ascending partial order).
+addMintValueToChangeMaps
+    :: (AssetId, TokenQuantity)
+    -> NonEmpty TokenMap
+    -> NonEmpty TokenMap
+addMintValueToChangeMaps (assetId, assetQty) = \case
+    -- This is the largest element, add mint value here
+    (x :| []) ->
+        TokenMap.add (TokenMap.singleton assetId assetQty) x :| []
+    -- There is a larget element, recurse to end of list
+    (x :| (y : rest)) ->
+        let
+            xs = y :| rest
+        in
+          x `NE.cons` addMintValueToChangeMaps (assetId, assetQty) xs
+
+addMintValuesToChangeMaps
+  :: TokenMap
+  -> NonEmpty TokenMap
+  -> NonEmpty TokenMap
+addMintValuesToChangeMaps mints =
+    F.foldr
+        (\mint -> (addMintValueToChangeMaps mint .))
+        id
+        (TokenMap.toFlatList mints)
+
+-- | Remove a burnt value from the change map. This function presumes that
+-- the change map is partially-ordered in ascending order (i.e. the largest
+-- element is at the end of the list). This function operates under the
+-- precondition that a valid selection has been performed, and so assumes
+-- that there is enough asset quantity in the change map to satisfy the
+-- burn. If there is not, it will burn as much as it can.
+--
+-- Properties:
+-- 1. wrt to the total value of assets in the changeMaps, calling this
+-- function is the same as removing the burnt value from the change maps:
+--   F.fold changeMaps `TokenMap.difference` TokenMap.singleton assetId qty
+--          == F.fold (f (assetId, qty) changeMaps)
+-- 2. length is conserved
+-- 3. order is preserved (i.e., the result is in ascending partial order).
+removeBurnValueFromChangeMaps
+    :: (AssetId, TokenQuantity)
+    -> NonEmpty TokenMap
+    -> NonEmpty TokenMap
+removeBurnValueFromChangeMaps (assetId, assetQty) =
+    burnWithGas assetQty
+  where
+      -- Subtract two token quantities, clamping negative values to zero.
+      subtract' :: TokenQuantity -> TokenQuantity -> TokenQuantity
+      subtract' x y =
+          fromMaybe TokenQuantity.zero (x `TokenQuantity.subtract` y)
```

Comment:
Suggestion: extract this into a `difference` function in the `TokenQuantity` module, and add a short property test for it.

For other types, it's common to use `difference` to capture a subtraction that's bounded by some type-specific concept of "zero". For example:

- [`TokenMap.difference`](https://github.com/input-output-hk/cardano-wallet/blob/fe50727c01b4b07f98bcfdd36be81a67e4b2d40b/lib/core/src/Cardano/Wallet/Primitive/Types/TokenMap.hs#L544)
- [`TokenBundle.difference`](https://github.com/input-output-hk/cardano-wallet/blob/fe50727c01b4b07f98bcfdd36be81a67e4b2d40b/lib/core/src/Cardano/Wallet/Primitive/Types/TokenBundle.hs#L322)
- [`Map.difference`](https://hackage.haskell.org/package/containers-0.6.4.1/docs/Data-Map-Strict.html#v:difference)
- [`Set.difference`](https://hackage.haskell.org/package/containers-0.6.4.1/docs/Data-Set.html#v:difference)

Bonus points: use this new function to simplify the definition of [`TokenMap.difference`](https://github.com/input-output-hk/cardano-wallet/blob/fe50727c01b4b07f98bcfdd36be81a67e4b2d40b/lib/core/src/Cardano/Wallet/Primitive/Types/TokenMap.hs#L544).

Reply:
Addressed in 89f100c40, bonus points in 5f839c7bf